pkg/syncer/resourcesync/controller.go

```diff
@@ -0,0 +1,330 @@
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), controllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) AddUpstreamEventHandler(handler ResourceEventHandlerPerGVR) {
+	c.upstreamEventHandlers = append(c.upstreamEventHandlers, handler)
+}
+
+func (c *Controller) AddDownstreamEventHandler(handler ResourceEventHandlerPerGVR) {
+	c.downstreamEventHandlers = append(c.downstreamEventHandlers, handler)
+}
+
+func (c *Controller) InformerForResource(gvr schema.GroupVersionResource) (*SyncerInformer, bool) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	if informer, ok := c.syncerInformerMap[gvr]; ok {
+		return informer, true
+	}
+
+	return nil, false
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("failed to sync %q: %w", key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	syncTarget, err := c.syncTargetLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		c.stopSyncerInformers(ctx, map[schema.GroupVersionResource]bool{})
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if syncTarget.GetUID() != c.syncTargetUID {
+		return nil
+	}
+
+	requiredGVRs := getAllGVRs(syncTarget)
+
+	for gvr := range requiredGVRs {
+		c.startSyncerInformer(ctx, gvr, syncTarget)
+	}
+
+	c.stopSyncerInformers(ctx, requiredGVRs)
+
+	return nil
+}
+
+// stopSyncer stop syncers for gvrs not in requiredGVRs
+func (c *Controller) stopSyncerInformers(ctx context.Context, requiredGVRs map[schema.GroupVersionResource]bool) {
+	logger := klog.FromContext(ctx)
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	for gvr, informer := range c.syncerInformerMap {
+		if _, ok := requiredGVRs[gvr]; !ok {
+			logger.WithValues("gvr", gvr.String()).V(2).Info("Stop syncer for gvr")
+			informer.cancel()
+			delete(c.syncerInformerMap, gvr)
+		}
+	}
+}
+
+func (c *Controller) startSyncerInformer(ctx context.Context, gvr schema.GroupVersionResource, syncTarget *workloadv1alpha1.SyncTarget) {
+	logger := klog.FromContext(ctx).WithValues("syncTarget", c.syncTargetName).WithValues("gvr", gvr.String())
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if _, ok := c.syncerInformerMap[gvr]; ok {
+		return
```

Comment:
how do you handle the case where the set of resources for a synctarget changes.?

Reply:
in sync target, we only track per gvr. If a gvr is included in the syncTarget, we start the informer when it is not started.